File: Hard-Array/Majortyelement3.java
import java.util.Scanner;

public class  Majortyelement3
{
    public static int majorityElement(int[] nums)
    {
        int count = 0, candidate = -1;

        for (int num : nums) {
            if (count == 0) {
                candidate = num;
                count = 1;
            } else if (num == candidate) {
                count++;
            } else {
                count--;
            }
        }

        // Verify the candidate
        count = 0;
        for (int num : nums) {
            if (num == candidate) {
                count++;
            }
        }

        return (count > nums.length / 3) ? candidate : -1;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the size of the array: ");
        int n = sc.nextInt();
        int[] arr = new int[n];
        
        System.out.println("Enter the elements of the array:");
        for (int i = 0; i < n; i++) 
        {
            arr[i] = sc.nextInt();
        }
        
        int result = majorityElement(arr);
        if (result == -1)
        {
            System.out.println("No Majority Element");
        }
        else
        {
            System.out.println("Majority Element is " + result);
        }
    }
    
}

File: Hard-Array/Majortyelement3.py
# WAP to find the majority element in an array. A majority element is an element that appears more than n/3 times in the array.
def majority_element(nums):
    count = {}
    for num in nums:
        if num in count:
            count[num] += 1
        else:
            count[num] = 1

    for num, freq in count.items():
        if freq > len(nums) // 3:
            return num

    return None

nums=int(input("Enter the number of elements in the array: "))
arr=[]
for i in range(nums):
    element=int(input("Enter the element: "))
    arr.append(element)
result = majority_element(arr)
print("The majority element is:", result)